### ISSUE TITLE
replace `cloud` with `ec2` (Proposed Breaking Change)

### ISSUE THREAD

**Issue by mrflip** [infochimps-labs/ironfan#66]
### The cloud statement needs rethunk

tl;dr -- The bare `cloud` statement (generic cloud attributes) was well-meaning but useless. 

Given that the only cloud we currently allow is EC2, I propose we remove the `cloud` directive (`cloud.flavor('t1.micro')`) in favor of `ec2` (`ec2.flavor('t1.micro')`). I'd like to not deprecate the term but remove it -- breaking any script that currently calls it. It's a simple regex replace and it will be unambiguous why the call failed.

If you don't like that idea, speak up now.

---

Longer version:

The `cloud` statement is intended to let me say "Here, friends, is the platonic ideal of an industrial-strength hadoop cluster, wheresoever you are.  Should you find yourself in Rackspace, apply that ideal on various components sized thusly; if instead, EC2, on components sized like this."

We use the terms that fog very nicely provides for describing aspects of a machine (flavor, image_name, etc), so that if you say `cloud.flavor('whatever')` the code to apply that directive is shared across providers.

The DSL looks like this:

``` ruby
  # no cloud specified
  cloud do
    flavor 'c1.medium'
    elastic_ip '123.45.67.89'
  end

  # an equivalent way of doing the above
  cloud.flavor 'c1.medium'
  cloud.elastic_ip '123.45.67.89'

  cloud(:ec2) do
    flavor 'c1.medium'
    elastic_ip '123.45.67.89'
   end

  cloud(:ec2).flavor 'c1.medium'         # (yuck)
  cloud(:ec2).elastic_ip '123.45.67.89'
```

The idea was that with a bare `cloud` statement you would say "Here's generic description of cloud shape", vs `cloud(:ec2)` defining "Here's specifics if the cluster is launched on EC2".

Now: most things on the left are generic across clouds. The rest are typically harmless even if the cloud doesn't handle them.

However! almost everything that you'd put on the _right_ of those is _not_ cloud-agnostic. Even things like 'elastic_ip' seem global but of course only exist in the cloud that owns it.

This shows that a generic 'cloud' statement doesn't make any sense, and while we have the chance to make breaking changes I'd like to delete it.

Instead, we define directives `ec2` (and so on for other cloud providers). Each provider's class inherits from cloud (and so can be decorated with things that only make sense on that cloud). 

``` ruby
  ec2.flavor 'c1.medium' 
  ec2.elastic_ip '123.45.67.89'

  rackspace.flavor     '1024MB'
  rackspace.image_name 'rs_maverick'

  vagrant do
    vfs_path '/disk2/vagrants/gibbon'
  end
```

Since everywhere we currently say `cloud` we really mean `ec2`, it's a simple regex-replace. So I'd like to not deprecate the term but remove it -- breaking any script that currently calls it. This will also help isolate the provider-specific stuff in the cluster_chef tools (though it's the cookbooks that need the real de-linting).

If you don't like the idea of breakage, speak up now.

**Comment by temujin9** [comment]
There's nothing that cloud itself will mean afterward, so we're not losing anything with depreciation. I would prefer that users with big installations have as little avoidable upgrade pain as possible. Throw a warning this cycle & pull it the next would be my vote.

**Comment by mrflip** [comment]
I'm actually finding there are some things that may make sense across platforms -- keypair, image name, knife bootstrap directives. Given that 90% of people will only run one cloud, it's not important to make them isolate the provider-specific attributes.

Other things could sensibly be abstracted -- rather than specifying 'm1.xlarge', I could ask for `:ram => 15, :cores => 4` or `storage => 1500` and get the cheapest machine that met or exceeded each constraint; the default of `:price => :smallest` would get me a t1.micro on EC2, a 256MB on Rackspace. Availability zones are another case.

So the existing syntax is fine. Its semantics are
- `cloud` with no predicate underlies all `cloud` blocks
- Stuff that is irrelevant to other providers is acceptable -- it will be ignored
- Stuff that is wrong in the context of other providers will presumably cause a downstream error -- it's your responsibility to overlay with provider-correct values.

```
cloud do
  # this makes sense everywhere
  image_name            'maverick'    

  # this will be ignored by many providers, and the value is non-portable;
  # but if you only run in one cloud there's no reason you can't put it here
  # (or override it, as we do below)
  public_ip             '1.2.3.4'     

  # This is harmless for the other clouds used below
  availability_zones   ['us-east-1d']
end

# these only apply to ec2 launches.
# `ec2` is sugar for `cloud(:ec2)`.
ec2 do
  spot_price_fraction   0.4
end

vagrant do
  private_ip            '10.10.6.5'
end

private_cloud.public_ip '4.3.2.1'

```

**Comment by mrflip** [comment]
The other thing I realized is that the most important multi-cloud deploy is 'cloud and vagrant': using local VMs to test your recipes.

Having overlays -- so that the set of vagrant-specific things is small and well-identified -- makes a ton of sense.

What's more, though, I may want to override not just cloud behavior but anything I'm applying to the facet/cluster/server.  This would imply `cloud` should just be a scope for "if this is my cloud context, _also_ apply this behavior". The code to ship around and merge the cloud object as its own thing is still kinda grody (though I cleaned up the resolve! behavior to where it's tolerable) which is its own smell.

I'm not going to implement that behavior -- the current implementation is compatible with that more interesting semantics -- but that's probably what we should look forward to.

Will leave this open until version_3 lands.

**Comment by temujin9** [comment]
+1: feels cleaner than the original proposal.

**Comment by pcn** [comment]
+1 agree that defining requirements is a good idea.  I still like the idea of using the ec2 semantics for now, but the future looks good.
